### DISCUSSION TITLE
Registration with mask

### DISCUSSION THREAD

**Discussion by samcoveney** [dipy/dipy#2508] [Registration] [answered]
Does DIPY provide the option to register images using a mask on the static image? I see that "dipy.align.imaffine.ParzenJointHistogram" has several methods that include arguments smask and mmask, suggesting that this functionality is present somewhere, but I cannot see how supply a mask to registration routines.

Background: I am processing cardiac data and reimplementing my current workflow into DIPY, as a sanity check. It is common to provide a Region Of Interest in the form of a mask to registration algorithms, so that only the pixels in this mask are used to calculated Mutual information.

I would be interested in helping provide this functionality, if it is not present. Although I suppose that Elastix already includes this, so it may be better just to use that software.

Thanks

**Comment by skoudoro** [comment, marked as answer]
Hi @samcoveney,

Not yet, this is on our TODO list. 

For inspiration, you can look at this PR https://github.com/dipy/dipy/pull/2179 which is in work in progress, or this issue https://github.com/dipy/dipy/issues/1969 which provide some temporary solution.

**Comment by skoudoro** [comment]
Also, any help is welcomed!  if you could test the PR #2179 and finalize it, it would be great!

**Comment by samcoveney** [comment]
Thanks for getting back to me so quickly, I'll take a look into this!

**Comment by samcoveney** [comment]
So I've forked #2179 and rebased, followed by making some fixes in the code and performing a test on a cardiac DWI registration problem using a static mask. I need to add some docstrings, and do a bunch of other checks (does this work moving image mask, can this be done with sparse sampling, etc.)

I can do some testing to make sure this works as intended, but should I provide code example in order for a pull request to get accepted? I could do with a hand on these final stages.

Thanks

**Comment by skoudoro** [comment]
Thank you very much for that. 

Testing, docstring is really important and this is our criterium to pass or not PR.

Concerning code examples, We highly recommend adding a tutorial or updating an existing one. However, we understand that it takes time to do a good tutorial so it can be done in another specific PR later.

Looking forward to testing your PR.

**Reply by samcoveney** [reply]
Affine registration without masking, fails because the squares in the corners pin the image:

![image](https://user-images.githubusercontent.com/19683485/152158520-a9a63b1d-bd5d-4617-b879-b46a45acd847.png)

Affine registration with static image masking to exclude the border in the static image that contains these squares:

![image](https://user-images.githubusercontent.com/19683485/152158579-7e779e6b-5014-4eea-b89c-245ff61e1d8e.png)

I've written this up as an example. Should I add this example to "doc/examples_index.rst" and "doc/examples/valid_examples.txt" and include everything in the pull request?

**Reply by skoudoro** [reply]
> Should I add this example to "doc/examples_index.rst" and "doc/examples/valid_examples.txt" and include everything in the pull request?
Yes

nice example, simple and explicit

**Comment by arokem** [comment]
+1 please do!

**Comment by samcoveney** [comment]
I am seeing an unusual problem and would appreciate a sanity check.

If I use two levels for registration, say [10000, 1000], then I get a good result. If I then use three levels [10000, 1000, 1] then I get a bad result, since it looks as if a reset has been done between levels, so that the previously found transformation did not follow through to the next level, but was instead reset (this is with starting_affine = None). This could have been a result changes to imaffine that I began with, or a result of my own changes, but I cannot see how these changes would have altered things so fundamentally from the original code.

Now of course I could follow this through even further and file a bug report, but in the spirit of collaboration, I was hoping somebody here could help me to double check that this fundamental problem is not occurring in the original dipy code?

**Reply by samcoveney** [reply]
I may have determined that there is a bug. The code in imaffine.py uses a for loop over levels which assumes that the final level will always have full resolution, even if this is not what is specified.

    # this has 1000 iterations at half resolution, and 1 at full resolution
    level_iters = [10000, 1000, 1]
    sigmas = [3.0, 1.0, 0.0]
    factors = [4, 2, 1]

    # this works better because there might be 1000 iterations at full resolution
    level_iters = [10000, 1000]
    sigmas = [3.0, 1.0]
    factors = [4, 2]

If I am right, then is this a bug? Granted, one would usually set the final level to be full resolution, but this behaviour was certainly not what I expected.

**Comment by skoudoro** [comment]
> The code in imaffine.py uses a for loop over levels which assumes that the final level will always have full resolution, even if this is not what is specified.

Yes, the last level is always full resolution. Where it is not specified? 

So we can see it as wrong documentation or a bug. I need to look at it, Can you point out where you saw the opposite information?

**Reply by samcoveney** [reply]
The example https://dipy.org/documentation/1.4.1./examples_built/affine_registration_3d/#example-affine-registration-3d does not specify that the final level is always full resolution. If the user supplies the factors [4,2] then they would assume that this final factor was the final resolution, but it is not, instead the full resolution is used which is confusing.

The docstring for AffineRegistration class in imaffine.py also does not document this behaviour.

It might be that the simplest fix is to enforce that the final factor level must be 1. Alternatively, the way that the class ScaleSpace in scalespace.py could be altered to not automatically store the original image as the final resolution image that will get used. I suppose it depends on intended use.

**Reply by skoudoro** [reply]
Indeed, make complete sense, good catch!

We need to clarify this point. I will create an issue from your last comment. right now, it can not be my priority, but I will come back to that when I finish all my other stuff

**Reply by samcoveney** [reply]
Sure, no rush, I am quite happy getting this masking stuff done without needing such a fix. Outside of speeding up testing, I think it is pretty standard to use the final level at full resolution, so this is probably a fairly minor problem overall.

**Comment by samcoveney** [comment]
I am now implementing the masks into the pipeline workflow by modifying dipy/align/_public.py

On lines 473 - 485 there seems to be some redundancy, whereby if the first function in the pipeline is 'center_of_mass' then this is used before the other transformations (e.g. translation). But since it is possible to use other transformations with starting_affine = "mass", would it be acceptable for me to remove this apparent redundancy, or am I missing something here?

I do not want to change too much code just to implement masking, in case I am not seeing a use case.

**Comment by skoudoro** [comment]
It seems that the goal is to initialize a starting affine when the user does not define any pipeline. Some beginner users just call this function with `moving` and `static` only as an argument. Also, it is convenient for tutorials.

**Reply by samcoveney** [reply]
That is a good point, if the pipeline only consisted of ['center_of_mass'] then it would be required to be implemented independently of other transformations, else it would not be possible to use such a pipeline. Thanks for the heads up, I knew I was missing something
